ddtrace/internal/_threads/lock.hpp

```diff
@@ -0,0 +1,400 @@
+RLock_release(RLock* self)
+{
+    if (self->_locked <= 0) {
+        PyErr_SetString(PyExc_RuntimeError, "Lock is not acquired");
+        return NULL;
+    }
+
+    self->_mutex->unlock();
+    self->_locked--;
+
+    Py_RETURN_NONE;
+}
+
+// ----------------------------------------------------------------------------
+static PyObject*
+RLock_locked(RLock* self)
+{
+    if (self->_locked > 0) {
+        Py_RETURN_TRUE;
+    }
+
+    Py_RETURN_FALSE;
+}
+
+// ----------------------------------------------------------------------------
+static PyObject*
+RLock_enter(RLock* self)
+{
+    AllowThreads _;
+
+    self->_mutex->lock();
+
+    self->_locked++;
+
+    Py_RETURN_NONE;
+}
+
+// ----------------------------------------------------------------------------
+static PyObject*
+RLock_exit(RLock* self, PyObject* Py_UNUSED(args), PyObject* Py_UNUSED(kwargs))
+{
+    // This method is called when the lock is used in a "with" statement
+    if (RLock_release(self) == NULL) {
+        return NULL; // Propagate any error from release
+    }
+
+    Py_RETURN_FALSE;
+}
+
+// ----------------------------------------------------------------------------
+static PyMethodDef RLock_methods[] = {
+    { "acquire",
+      (PyCFunction)RLock_acquire,
+      METH_VARARGS | METH_KEYWORDS,
+      "Acquire the lock with an optional timeout" },
+    { "release", (PyCFunction)RLock_release, METH_NOARGS, "Release the lock" },
+    { "locked", (PyCFunction)RLock_locked, METH_NOARGS, "Return whether the lock is acquired at least once" },
+    { "__enter__", (PyCFunction)RLock_enter, METH_NOARGS, "Enter the lock context" },
+    { "__exit__", (PyCFunction)RLock_exit, METH_VARARGS | METH_KEYWORDS, "Exit the lock context" },
+    { NULL } /* Sentinel */
+};
+
+// ----------------------------------------------------------------------------
+static PyMemberDef RLock_members[] = {
+    { NULL } /* Sentinel */
+};
+
+// ----------------------------------------------------------------------------
+static PyTypeObject RLockType = {
+    .ob_base = PyVarObject_HEAD_INIT(NULL, 0).tp_name = "ddtrace.internal._threads.RLock",
+    .tp_basicsize = sizeof(RLock),
+    .tp_itemsize = 0,
+    .tp_dealloc = (destructor)RLock_dealloc,
+    .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE,
+    .tp_doc = PyDoc_STR("Native re-entrant lock implementation"),
+    .tp_methods = RLock_methods,
+    .tp_members = RLock_members,
+    .tp_init = (initproc)RLock_init,
+    .tp_new = PyType_GenericNew,
+};
+
+// ----------------------------------------------------------------------------
+static PyObject*
+lock_acquire_all(PyObject* Py_UNUSED(self), PyObject* Py_UNUSED(args))
+{
+    {
+        AllowThreads _;
+
+        _lock_set_mutex.lock();
+
+        for (Lock* lock : lock_set) {
+            lock->_mutex->lock();
+            lock->_locked = 1;
+        }
+
+        for (RLock* rlock : rlock_set) {
+            rlock->_mutex->lock();
+            rlock->_locked++;
+        }
```

Comment:
How can we sure that this way of acquiring all locks in non-deterministic order doesn't result in a deadlock?

I was wondering what if there is one thread holding lock A and waiting for lock B, while another thread tries to fork() acquires B and try to acquire A.